493 show rate limiting error message to the user

pmarsh-scottlogic opened the pull request:
## Description

when we get rate limited, we see that in the front end in a simple error message to the user. When we get any other error from chatGPT we see that error in its original form.

## Screenshots

![image](https://github.com/ScottLogic/prompt-injection/assets/118171430/b8c68691-9979-4871-aa83-a9399c1be344)

## Notes

- small refactor to shorten a try-catch block.
- adds new method `handleErrorGettingReply` for repeated code - adding error message to chat history and sending error response

## Concerns

- I'm not sure what other errors will be thrown by the openAI API or how to recreate them. Right now now the error message in full is shown to the user, which is a bit intimidating I imagine. Could change that back to a generic "Failed to get GPT reply"

## Checklist

Have you done the following?

- [x] Linked the relevant Issue
- [x] Added tests
- [x] Ensured the workflow steps are passing

pmarsh-scottlogic commented:
to recreate, you can use this free api key `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`

heatherlogan-scottlogic commented on backend/src/controller/chatController.ts:
```diff
@@ -187,38 +192,79 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
+		);
+		return;
+	} else if (!chatResponse.reply || chatResponse.reply === '') {
+		handleErrorGettingReply(
```
i think there shouldn't be an error here as we are have responses which are blocked by defences set to null. treating as an error then results in 

![image](https://github.com/ScottLogic/prompt-injection/assets/118981273/d1ae4c08-c264-4ba4-b3de-beb2877fc5e6) (char limit set to 20)

pmarsh-scottlogic replied:
good spot! Fixed

pmarsh-scottlogic replied:
I'll add a unit test for this case too

gsproston-scottlogic commented on backend/src/controller/chatController.ts:
```diff
@@ -187,38 +192,79 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
+			simplifyOpenAIErrorMessage(chatResponse.openAIErrorMessage)
+		);
+		return;
+	} else if (!chatResponse.reply || chatResponse.reply === '') {
```
Empty strings are falsy, so we don't need the ` || chatResponse.reply === ''`

pmarsh-scottlogic commented on backend/src/controller/chatController.ts:
```diff
@@ -187,38 +192,79 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
+	if (openAIErrorMessage.startsWith('429')) {
+		const tryAgainMessage = openAIErrorMessage
+			.split('. ')
+			.find((sentence) => sentence.includes('Please try again in'));
```
oh, um. I actually didn't check this. That was my assumption. Let me investigate...

pmarsh-scottlogic replied:
Ok. I know that the timeout changes. [Here's an example](https://community.openai.com/t/rate-limit-reached-for-10ktpm-200rpm/381878) of a different timeout. I suppose it's different per membership. But also, because there is a rate limit per minute distinct from the rate limit per day, I expect that the timeouts will reflect these. i.e. if you've exceeded the limit per day it would say (I hope) "please try again tomorrow"

heatherlogan-scottlogic replied:
Ah cool, thanks for confirming!

gsproston-scottlogic reviewed (approved):
Just one thing from me, aside from Heather's comments. It's not blocking approval though.

The key unfortunately doesn't work for me. It fails when vectorising the documents, and the backend crashes. But this all looks sensible.

pmarsh-scottlogic commented:
> Just one thing from me, aside from Heather's comments. It's not blocking approval though.
> 
> The key unfortunately doesn't work for me. It fails when vectorising the documents, and the backend crashes. But this all looks sensible.

Hmm yeah, my free keys suddenly started expiring very quickly. Ah well, guess we'll have to generate our own when we need 'em

pmarsh-scottlogic commented:
I've added some tests to make sure the controller behaves well when a defence is blocked. I also moved some of the controller tests to the integration test folder because I realised they rely heavily on code in `openAI.ts`. Happy to undo though, I know the distinction is blurry sometimes

heatherlogan-scottlogic reviewed (approved):
great stuff ⭐
